compat/socket.cc

```diff
@@ -0,0 +1,198 @@
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = bind(handle,n,l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xconnect(int socketFd, const struct sockaddr * n, socklen_t l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = connect(handle,n,l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xgetsockname(int sockfd, struct sockaddr * addr, socklen_t * addrlen)
+{
+    const auto handle = _get_osfhandle(sockfd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    auto al = static_cast<int>(*addrlen);
+    const auto result = getsockname(handle, addr, &al);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    *addrlen = static_cast<socklen_t>(al);
+    return result;
+}
+
+int
+xgetsockopt(int socket, int level, int option_name, void * option_value, socklen_t * option_len)
+{
+    const auto handle = _get_osfhandle(socket);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    auto ol = static_cast<int>(*option_len);
+    const auto result = getsockopt(handle, level, option_name, static_cast<char *>(option_value), &ol);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    *option_len = static_cast<socklen_t>(ol);
+    return result;
+}
+
+int
+xlisten(int socketFd, int backlog)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = listen(handle, backlog);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+ssize_t
+xrecv(int socketFd, void * buf, size_t len, int flags)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    assert(len <= INT_MAX);
+    const auto result = recv(handle, static_cast<char *>(buf), static_cast<int>(len), flags);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+ssize_t
+xrecvfrom(int socketFd, void * buf, size_t len, int flags, struct sockaddr * from, socklen_t * fromlen)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (!isValidSocketHandle(handle)) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    auto fl = static_cast<int>(*fromlen);
```

Comment:
WIndows and Linux documentation says that fromlen may be nil. Please do not dereference without checking.

While fixing this, please rename `fromlen` to `fromLen` or similar to use camelCase and make this variable more readable.

Reply:
done by making fl a pointer and passing it through to the  library

Reply:
reworked in 2605019d1f